transformer_engine/pytorch/attention.py

```diff
@@ -1031,23 +1040,58 @@ def forward(self, max_seq_len: int, offset: int = 0):
 
 
 class FusedRoPEFunc(torch.autograd.Function):
+    """
+    Function for FusedRoPE
+
+    This implementation assumes the input tensor to be in `sbhd`, `bshd` or `thd` format and
+    the RoPE tensor to be of shape (s, 1, 1, d). It accepts arbitrary memory layouts to avoid
+    the expensive `.contiguous()` calls, thus it may not achieve the best memory access pattern.
+    """
+
     @staticmethod
     def forward(
-        ctx, t: torch.Tensor, cos_: torch.Tensor, sin_: torch.Tensor
+        ctx,
+        t: torch.Tensor,
+        freqs: torch.Tensor,
+        transpose_output_memory: bool = False,
+        qkv_format: str = "sbhd",
```

Comment:
Typically we would expect the output format to match the input format, so the `transpose_output_memory` kwarg seems to make the API more complicated than necessary. Could we remove `transpose_output_memory` as a kwarg and internally calculate based on `qkv_format == "bshd"`?

Reply:
Removed `transpose_output_memory` and add a comment to explain the behavior.

Reply:
I don't think this is the right API since it still forces the output to be SBHD. It's unintuitive for the user if they put in a BSHD input.